internal/util/diagtree/diagtree.go

```diff
@@ -70,16 +70,36 @@ func (m *Node) levelPrefix(level int) string {
 	return strings.Repeat("  ", (level-2)*2) + "- "
 }
 
+type cappedWriter struct {
+	max int
```

Comment:
Nit: If you're mutating this rather than having a seprate counter I'd suggest naming it something like `remaining` to indicate it's not the original max value.